chain/processor_test.go

```diff
@@ -0,0 +1,618 @@
+				require.NoError(t, db.Put([]byte(heightKey), binary.BigEndian.AppendUint64(nil, 0)))
+				require.NoError(t, db.Put([]byte(timestampKey), binary.BigEndian.AppendUint64(nil, 0)))
+				require.NoError(t, db.Put([]byte(feeKey), []byte{}))
+				return db
+			}(),
+			newBlockF: func(ids.ID) *chain.StatelessBlock {
+				block, err := chain.NewStatelessBlock(
+					ids.Empty,
+					genesis.NewDefaultRules().GetMinEmptyBlockGap(),
+					1,
+					nil,
+					ids.GenerateTestID(),
+				)
+				require.NoError(t, err)
+				return block
+			},
+			expectedErr: chain.ErrStateRootMismatch,
+		},
+		{
+			name:           "failed to verify signatures",
+			validityWindow: &validitywindowtest.MockTimeValidityWindow[*chain.Transaction]{},
+			workers:        workers.NewSerial(),
+			db: func() merkledb.MerkleDB {
+				db, err := merkledb.New(
+					context.Background(),
+					memdb.New(),
+					merkledb.Config{
+						BranchFactor: merkledb.BranchFactor16,
+						Tracer:       trace.Noop,
+					},
+				)
+				require.NoError(t, err)
+				require.NoError(t, db.Put([]byte(heightKey), binary.BigEndian.AppendUint64(nil, 0)))
+				require.NoError(t, db.Put([]byte(timestampKey), binary.BigEndian.AppendUint64(nil, 0)))
+				require.NoError(t, db.Put([]byte(feeKey), []byte{}))
+				return db
+			}(),
+			newBlockF: func(parentRoot ids.ID) *chain.StatelessBlock {
+				block, err := chain.NewStatelessBlock(
+					ids.Empty,
+					genesis.NewDefaultRules().GetMinEmptyBlockGap(),
+					1,
+					[]*chain.Transaction{
+						func() *chain.Transaction {
+							p, err := ed25519.GeneratePrivateKey()
+							require.NoError(t, err)
+
+							testRules := genesis.NewDefaultRules()
+							tx, err := chain.NewTransaction(
+								&chain.Base{
+									Timestamp: utils.UnixRMilli(
+										genesis.NewDefaultRules().GetMinEmptyBlockGap(),
+										testRules.GetValidityWindow(),
+									),
+								},
+								[]chain.Action{},
+								&auth.ED25519{
+									Signer: p.PublicKey(),
+								},
+							)
+							require.NoError(t, err)
+							return tx
+						}(),
+					},
+					parentRoot,
+				)
+				require.NoError(t, err)
+				return block
+			},
+			expectedErr: crypto.ErrInvalidSignature,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := require.New(t)
+			ctx := context.Background()
+
+			metrics, err := chain.NewMetrics(prometheus.NewRegistry())
+			r.NoError(err)
+
+			processor := chain.NewProcessor(
+				trace.Noop,
+				&logging.NoLog{},
+				&genesis.ImmutableRuleFactory{Rules: genesis.NewDefaultRules()},
+				tt.workers,
+				&mockAuthVM{},
+				metadata.NewDefaultManager(),
+				&mockBalanceHandler{},
+				tt.validityWindow,
+				metrics,
+				chain.NewDefaultConfig(),
+			)
+
+			root, err := tt.db.GetMerkleRoot(ctx)
+			r.NoError(err)
+
+			_, err = processor.Execute(
+				ctx,
+				tt.db,
+				chain.NewExecutionBlock(tt.newBlockF(root)),
+				tt.isNormalOp,
+			)
+			r.ErrorIs(err, tt.expectedErr)
```

Comment:
(separate PR) - could we confirm the post state is what we expect as well ie. we should make sure that the metadata is updated correctly and that tx/action execution (include successful and failed w/ rollbacks) has the expected effect